### PR TITLE
Add tests for core PyTorch files

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Add unit tests and fix bugs/add comments as necessary to PyTorch core files (`networks.py`, `encoders.py`, `decoders.py`, `utils.py`)

Tests completed:
- [x] `utils.py`
- [x] `encoders.py`
- [x] `decoders.py`
- [x] `networks.py`

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/tests/torch/test_decoders.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    # Test default 1 value per head
+    value_heads = ValueHeads(stream_names, input_size)
+    input_data = torch.ones((batch_size, input_size))
+    value_out, _ = value_heads(input_data)  # Note: mean value will be removed shortly
+
+    for stream_name in stream_names:
+        assert value_out[stream_name].shape == (batch_size,)
+
+    # Test that iinputting the wrong size input will throw an error
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/torch/test_decoders.py:20]:
```suggestion
    # Test that inputting the wrong size input will throw an error
```

File: ml-agents/mlagents/trainers/tests/torch/test_distribution.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            test_prob = torch.tensor(
+                [1.0 - 0.01 * (act_size[i] - 1)] + [0.01] * (act_size[i] - 1)
+            )  # High prob for first action
+            test_log_prob = test_prob.log()
+            loss += torch.nn.functional.mse_loss(log_prob, test_log_prob)
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+    for dist_inst in dist_insts:
+        assert dist_inst.all_log_prob().flatten()[0] == pytest.approx(0, abs=0.1)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/torch/test_distribution.py:69]: Why are the log probs supposed to be 0 here ? it is unclear to me given the loss

File: ml-agents/mlagents/trainers/tests/torch/test_distribution.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    for branch in act_size:
+        masks += [0] * (branch - 1) + [1]
+    masks = torch.tensor([masks])
+    dist_insts = gauss_dist(sample_embedding, masks=masks)
+    for dist_inst in dist_insts:
+        log_prob = dist_inst.all_log_prob()
+        assert log_prob.flatten()[-1] == pytest.approx(0, abs=0.001)
+
+
+def test_gaussian_dist_instance():
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/torch/test_distribution.py:82]: Are we setting the seed somewhere?

File: ml-agents/mlagents/trainers/tests/torch/test_distribution.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        [1.0 - 0.1 * (act_size - 1)] + [0.1] * (act_size - 1)
+    )  # High prob for first action
+    dist_instance = CategoricalDistInstance(test_prob)
+
+    for _ in range(10):
+        action = dist_instance.sample()
+        assert action.shape == (1,)
+        assert action < act_size
+
+    # Make sure log_prob of 1st action is high
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/torch/test_distribution.py:121]: This comment says "make sure", but there is no assert here.

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/tests/torch/test_distribution.py:121]: I would combine the two comments into : Make sure the first action has higher probability than the others

File: ml-agents/mlagents/trainers/tests/torch/test_encoders.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+
+# This test will also reveal issues with states not being saved in the state_dict.
+def compare_models(module_1, module_2):
+    is_same = True
+    for key_item_1, key_item_2 in zip(
+        module_1.state_dict().items(), module_2.state_dict().items()
+    ):
+        is_same = torch.equal(key_item_1[1], key_item_2[1]) and is_same
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/torch/test_encoders.py:21]: Why compare only the first item? `key_item_1[*1*]`?

File: ml-agents/mlagents/trainers/torch/encoders.py [hunk 6]
```diff
@@ -9,9 +9,9 @@
 class Normalizer(nn.Module):
     def __init__(self, vec_obs_size: int):
         super().__init__()
-        self.normalization_steps = torch.tensor(1)
-        self.running_mean = torch.zeros(vec_obs_size)
-        self.running_variance = torch.ones(vec_obs_size)
+        self.register_buffer("normalization_steps", torch.tensor(1))
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/encoders.py:12]: I think we need to determine if we want to use `Parameter(require_grad=False)` or `register_buffer`. I tend to prefer the first one since it is more intuitive what it means. I think @dongruoping is also using `Parameter` in her PR.

Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/encoders.py:12]: I do agree Parameter is more intuitive. Went with `register_buffer` b/c it's what's used by PyTorch in their own code. E.g. running mean and variance in the Batch Norm module: https://pytorch.org/docs/stable/_modules/torch/nn/modules/batchnorm.html#BatchNorm1d

I agree though we should make it the same across all our code. @dongruoping what do you think?

File: ml-agents/mlagents/trainers/tests/torch/test_distribution.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/tests/torch/test_distribution.py:82]: Added seed sets.

File: ml-agents/mlagents/trainers/tests/torch/test_encoders.py (hunk 5 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/tests/torch/test_encoders.py:21]: It's the ZIP, `key_item_X` is a Tuple of (key, value). So we're just comparing the values.

File: ml-agents/mlagents/trainers/tests/torch/test_distribution.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/tests/torch/test_distribution.py:69]: It's checking that the probability of the 1st action is near 1 (which means the log prob is 0). I suppose it would be more clear to just check if it matches the entire test_log_prob, so I've changed it to do that.

File: ml-agents/mlagents/trainers/torch/encoders.py (hunk 6 above)
Comment (dongruoping) [reply ml-agents/mlagents/trainers/torch/encoders.py:12]: They're pretty much the same effectively but I think Pytorch made this buffer thing to avoid the impression of this variable being something "trainable" and also avoid forgetting to set require_grad=False. The other benefit is that these will not need be passed to the optimizer and then be skipped by optimizer when updating. Not big issue.
The name `buffer` is not the best though I agree.
I'm good with both as long as we all use the same one.

Comment (dongruoping) [reply ml-agents/mlagents/trainers/torch/encoders.py:12]: Actually if it's used by PyTorch maybe it would be good to follow.
